Add wp.org SVN automation, mainly copied from Genesis Blocks

kienstra opened the pull request:
<!--- Please summarize this PR in the title above -->

#### Changes
* Adds a wp.org SVN deployment step, thanks to [Genesis Blocks](https://github.com/studiopress/genesis-blocks/blob/37c4a96a8c8563bb0df62959afbeb33f364c279f/.circleci/config.yml#L43)
* On creating a tag like `1.0.2`, this deploys to wp.org
* It won't deploy for tags like [1.0.1-built](https://github.com/studiopress/genesis-custom-blocks/releases/tag/1.0.1-built), which is a tag for Composer or git submodule use

#### Testing instructions
1. `npm ci && npm run gulp`
1. Upload and activate `package/genesis-custom-blocks.1.0.1.zip` to another WP instance
1. Expected: it activates without error, and the submenus are there, like 'All Blocks', 'Add New', 'Settings', and 'Genesis Pro'
1. `circleci local execute --job svn-deploy`
1. Expected: the stat that is shown at the end is something like:
<details>
<summary>Expected stat</summary>

```
Here is the svn stat about to be checked in:
?       genesis-custom-blocks.1.0.1.zip
M       tags/1.0.1/vendor/autoload.php
M       tags/1.0.1/vendor/composer/autoload_real.php
M       tags/1.0.1/vendor/composer/autoload_static.php
M       trunk/vendor/autoload.php
M       trunk/vendor/composer/autoload_real.php
M       trunk/vendor/composer/autoload_static.php
Success!
```
</details>

6. Simulate a new version:
```diff
diff --git a/CHANGELOG.md b/CHANGELOG.md
index f97d468..c060ff9 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,3 +14,7 @@ Fix an error if Block Lab 1.5.6 is also active
 
 * Fixes an error with Block Lab 1.5.6, where it defines functions twice
 * Error does not occur with latest Block Lab
+
+### 1.0.2 - 2020-09-13 ###
+
+Fix a bug
diff --git a/README.md b/README.md
index 1aacc92..b6b8bcd 100644
--- a/README.md
+++ b/README.md
@@ -5,7 +5,7 @@ Tags: gutenberg, blocks, block editor, fields, template
 Requires at least: 5.0
 Tested up to: 5.5
 Requires PHP: 5.6
-Stable tag: 1.0.1
+Stable tag: 1.0.2
 License: GPLv2 or later
 License URI: http://www.gnu.org/licenses/gpl
 
diff --git a/genesis-custom-blocks.php b/genesis-custom-blocks.php
index c10c48a..6467cda 100644
--- a/genesis-custom-blocks.php
+++ b/genesis-custom-blocks.php
@@ -8,7 +8,7 @@
  *
  * Plugin Name: Genesis Custom Blocks
  * Description: The easy way to build custom blocks for Gutenberg.
- * Version: 1.0.1
+ * Version: 1.0.2
  * Author: Genesis Custom Blocks
  * Author URI: https://studiopress.com
  * License: GPL2
diff --git a/package.json b/package.json
index 3ed2035..0b5615d 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "genesis-custom-blocks",
   "title": "Genesis Custom Blocks",
-  "version": "1.0.1",
+  "version": "1.0.2",
   "description": "WordPress plugin with a simple templating system for building custom blocks.",
   "author": "Genesis Custom Blocks",
   "license": "GPL-2.0-or-later",
```
7. `circleci local execute --job svn-deploy`

8. Expected: the stat that the job shows at the end is something like:
<details>
<summary>Expected stat</summary>

```
Here is the svn stat about to be checked in:
?       genesis-custom-blocks.1.0.2.zip
A       tags/1.0.2
A       tags/1.0.2/LICENSE
A       tags/1.0.2/assets
A       tags/1.0.2/assets/icons.json
A       tags/1.0.2/assets/images
A       tags/1.0.2/assets/images/admin-menu-icon.svg
A       tags/1.0.2/assets/images/logo-reversed.svg
A       tags/1.0.2/css
A       tags/1.0.2/css/admin.block-edit.css
A       tags/1.0.2/css/admin.block-post.css
A       tags/1.0.2/css/admin.css
A       tags/1.0.2/css/admin.onboarding.css
A       tags/1.0.2/css/admin.upgrade.css
A       tags/1.0.2/css/blocks.editor.asset.php
A       tags/1.0.2/css/blocks.editor.css
A       tags/1.0.2/genesis-custom-blocks.php
A       tags/1.0.2/js
A       tags/1.0.2/js/admin.block-post.js
A       tags/1.0.2/js/editor.blocks.asset.php
A       tags/1.0.2/js/editor.blocks.js
A       tags/1.0.2/language
A       tags/1.0.2/php
A       tags/1.0.2/php/Admin
A       tags/1.0.2/php/Admin/Admin.php
A       tags/1.0.2/php/Admin/Documentation.php
A       tags/1.0.2/php/Admin/Import.php
A       tags/1.0.2/php/Admin/Onboarding.php
A       tags/1.0.2/php/Admin/Upgrade.php
A       tags/1.0.2/php/BlockApi.php
A       tags/1.0.2/php/Blocks
A       tags/1.0.2/php/Blocks/Block.php
A       tags/1.0.2/php/Blocks/Controls
A       tags/1.0.2/php/Blocks/Controls/Checkbox.php
A       tags/1.0.2/php/Blocks/Controls/Color.php
A       tags/1.0.2/php/Blocks/Controls/ControlAbstract.php
A       tags/1.0.2/php/Blocks/Controls/ControlSetting.php
A       tags/1.0.2/php/Blocks/Controls/Email.php
A       tags/1.0.2/php/Blocks/Controls/Image.php
A       tags/1.0.2/php/Blocks/Controls/Multiselect.php
A       tags/1.0.2/php/Blocks/Controls/Number.php
A       tags/1.0.2/php/Blocks/Controls/Radio.php
A       tags/1.0.2/php/Blocks/Controls/Range.php
A       tags/1.0.2/php/Blocks/Controls/Select.php
A       tags/1.0.2/php/Blocks/Controls/Text.php
A       tags/1.0.2/php/Blocks/Controls/Textarea.php
A       tags/1.0.2/php/Blocks/Controls/Toggle.php
A       tags/1.0.2/php/Blocks/Controls/Url.php
A       tags/1.0.2/php/Blocks/Field.php
A       tags/1.0.2/php/Blocks/Loader.php
A       tags/1.0.2/php/ComponentAbstract.php
A       tags/1.0.2/php/ComponentInterface.php
A       tags/1.0.2/php/Deprecated.php
A       tags/1.0.2/php/Helpers.php
A       tags/1.0.2/php/Plugin.php
A       tags/1.0.2/php/PluginAbstract.php
A       tags/1.0.2/php/PluginInterface.php
A       tags/1.0.2/php/PostTypes
A       tags/1.0.2/php/PostTypes/BlockPost.php
A       tags/1.0.2/php/Util.php
A       tags/1.0.2/php/Views
A       tags/1.0.2/php/Views/Upgrade.php
A       tags/1.0.2/readme.txt
A       tags/1.0.2/vendor
A       tags/1.0.2/vendor/autoload.php
A       tags/1.0.2/vendor/composer
A       tags/1.0.2/vendor/composer/ClassLoader.php
A       tags/1.0.2/vendor/composer/LICENSE
A       tags/1.0.2/vendor/composer/autoload_classmap.php
A       tags/1.0.2/vendor/composer/autoload_namespaces.php
A       tags/1.0.2/vendor/composer/autoload_psr4.php
A       tags/1.0.2/vendor/composer/autoload_real.php
A       tags/1.0.2/vendor/composer/autoload_static.php
A       tags/1.0.2/vendor/composer/installed.json
M       trunk/genesis-custom-blocks.php
M       trunk/readme.txt
M       trunk/vendor/autoload.php
M       trunk/vendor/composer/autoload_real.php
M       trunk/vendor/composer/autoload_static.php
Success!
```

</details>

9. `touch php/Foo.php && rm php/Plugin.php`
10. `git add php/Foo.php php/Plugin.php && git commit --no-verify -m "Will revert: test adding and deleting a file"`
11. `circleci local execute --job svn-deploy`
12. Expected: the stat is something like:
<details>
<summary>Expected stat</summary>

```
Here is the svn stat about to be checked in:
?       genesis-custom-blocks.1.0.1.zip
A       tags/1.0.1/php/Foo.php
D       tags/1.0.1/php/Plugin.php
M       tags/1.0.1/vendor/autoload.php
M       tags/1.0.1/vendor/composer/autoload_classmap.php
M       tags/1.0.1/vendor/composer/autoload_real.php
M       tags/1.0.1/vendor/composer/autoload_static.php
A       trunk/php/Foo.php
D       trunk/php/Plugin.php
M       trunk/vendor/autoload.php
M       trunk/vendor/composer/autoload_classmap.php
M       trunk/vendor/composer/autoload_real.php
M       trunk/vendor/composer/autoload_static.php
Success!
```
</details>

13. We normally won't be deploying without bumping the version, but that tests that files were deleted and added as expected.

kienstra committed:
Don't add assets to build process

kienstra committed:
Mainly copy CircleCI wp.org deploy job from Genesis Blocks

kienstra committed:
Add set-version step, mainly taken from Genesis Blocks again

kienstra committed:
Restore building of wp.org assets like screenshots

kienstra committed:
Prevent removing the tags/ directory, as there will be other tags

kienstra committed:
Possible change

kienstra committed:
Use svn up as the directories are already under version control

svn add casued an error,
as tags/, trunk/, and assets/
were already under version control.

kienstra committed:
Add new files that weren't added before

There can be files
that weren't added before,
and they needed to be added.

kienstra committed:
Remove the svn rm statement, as it probably won't be used

Right now, there's no way to
remove files from trunk/, assets/, or tags/

mindctrl commented on .circleci/config.yml:
```diff
@@ -111,3 +137,7 @@ workflows:
+  deploy:
+    jobs:
+      - svn-deploy:
+          context: genesis-svn
```
@kienstra you'll want something in here so that it only deploys when tags are pushed. In AB we do this:

```
          filters:
            tags:
              only: /^\d+\.\d+\.\d+$/
            branches:
              ignore: /.*/
```

https://github.com/studiopress/atomic-blocks/blob/0d1808b57bcb4caf99ce6de419be1876fac7d02f/.circleci/config.yml#L280-L288

kienstra replied:
Thanks, good idea!

kienstra replied:
Applied in https://github.com/studiopress/genesis-custom-blocks/pull/32/commits/1a5eca83a7a8bd3e309c6ef9cf765beec26a204e

kienstra committed:
Only run svn-deploy on tags, thanks to John's suggestion

This only runs on tags like
1.0.2, but not on 1.0.2-built

kienstra committed:
Add new files, and remove deleted ones

In SVN, it's necessary to
rm files from version control.

kienstra committed:
Store the SVN directories in a variable, modernize gulpfile.js

It was written when const
wasn't used as much.

kienstra committed:
Remove extra set-version step, not used anymore

kienstra commented on .circleci/config.yml:
```diff
@@ -99,6 +99,28 @@ jobs:
+            npm ci && npm run gulp
+            cd package
+            SVN_DIRECTORIES="trunk tags assets"
+            svn stat $SVN_DIRECTORIES | { grep -E '^\?' || true; } | awk '{print $2}' | xargs -r svn add
```
Adds new files to version control, mainly in `tags/<new version number>/*`

kienstra commented on .circleci/config.yml:
```diff
@@ -99,6 +99,28 @@ jobs:
+            cd package
+            SVN_DIRECTORIES="trunk tags assets"
+            svn stat $SVN_DIRECTORIES | { grep -E '^\?' || true; } | awk '{print $2}' | xargs -r svn add
+            svn stat $SVN_DIRECTORIES | { grep -E '^\!' || true; } | awk '{print $2}' | xargs -r svn rm
```
Removes files from version control if we've deleted them. Probably not common, but we might delete or rename JS files.

kienstra commented on gulpfile.js:
```diff
@@ -1,32 +1,27 @@
-var gulp = require( 'gulp' );
-var merge = require('merge-stream');
```
`merge` wasn't used

kienstra committed:
Remove the needles source statement

this used to get an environmental
variable, and needed this.

kienstra commented on .circleci/config.yml:
```diff
@@ -99,6 +99,28 @@ jobs:
+            svn up $SVN_DIRECTORIES
+            echo "Here is the svn stat about to be checked in:"
+            svn stat
+
```
I'll add the `svn ci` statement here after a review, I just don't want any of the local testing to accidentally push. Even though that's probably not possible.

```bash
svn ci -m "Tagging ${BUILD_VERSION} from GitHub" --no-auth-cache --non-interactive --username "${SVN_USERNAME}" --password "${SVN_PASSWORD}"
```

mindctrl replied:
👍  command looks good to me.

kienstra commented on gulpfile.js:
```diff
@@ -1,32 +1,27 @@
-
-gulp.task( 'version', function () {
-	var pluginStream = gulp.src( [ 'genesis-custom-blocks.php' ] )
-		.pipe( replace( new RegExp( /Version:\s*(.*)/, 'g' ), "Version: " + config.version ) )
```
This shouldn't be needed.

We're manually bumping the `Version` each release:

https://github.com/studiopress/genesis-custom-blocks/blob/5636c1e522886286dcade596697d16ab69be510e/genesis-custom-blocks.php#L11

kienstra committed:
Add a BUILD_VERSION variable to use in the svn ci message

kienstra commented:
Hi @mindctrl and @nickcernis,
Could you please review this when you have a chance? 

Thanks!

mindctrl commented on .circleci/config.yml:
```diff
@@ -111,3 +133,12 @@ workflows:
+            tags:
+              only: /^\d+\.\d+\.\d+$/
+            branches:
+              ignore: /.*/
```
`filters` looks good now.

Another thing you might want here is a `requires` configuration, so that the deploy doesn't happen if the tests fail. You can see how we do that in AB [here](https://github.com/studiopress/atomic-blocks/blob/0d1808b57bcb4caf99ce6de419be1876fac7d02f/.circleci/config.yml#L282-L283).

kienstra replied:
Oh, good idea!

kienstra replied:
Applied in https://github.com/studiopress/genesis-custom-blocks/pull/32/commits/dcbbef04005ab668d51040ebbddd759f69f359e2

kienstra committed:
Require linting and PHPUnit tests before SVN deployment

Thanks to John for the suggestion
This should make it more reliable.

kienstra committed:
Update CONTRIBUTING.md, mention that CircleCI deploys to wp.org

Also, update tag-built.sh
so that it does npm ci
and npm run gulp.
Also, minor changes to tag-built.sh
so that the tags are available.

mindctrl commented on .circleci/config.yml:
```diff
@@ -99,15 +99,45 @@ jobs:
+          command: |
+            BUILD_VERSION=$(grep 'Version:' genesis-custom-blocks.php | cut -f4 -d' ')
+            sudo apt-get update && sudo apt-get install subversion
+            svn co https://plugins.svn.wordpress.org/genesis-custom-blocks package
```
One thing [we do in AB](https://github.com/studiopress/atomic-blocks/blob/0d1808b57bcb4caf99ce6de419be1876fac7d02f/.circleci/config.yml#L47) is use the `--depth=empty` option when doing a checkout. The reason was the amount of time it takes to do a checkout once you have a sufficient number of releases out there. The svn server is just so slow sometimes. Another option is to prune tags every now and then, which [wp.org recommends](https://developer.wordpress.org/plugins/wordpress-org/plugin-developer-faq/#how-many-old-releases-should-i-keep-in-svn) - we just never got around to implementing that.

Not a blocker, but just mentioning for future consideration.

kienstra replied:
Good idea, I'll make a commit for that

kienstra replied:
Hm, maybe pruning would be good when there are several tags.

When adding `--depth=empty` to `svn co`, `circleci local execute --job svn-deploy` shows this message:

```
Searching tree conflict details for 'assets' in repository:
Checking r2372653... done
Tree conflict on 'assets':
A new directory appeared during update to r2378985; it was added by plugin-master in r2372653.
A directory scheduled to be added to the repository in the next commit was found in the working copy.
Select: (p) Postpone, (r) Mark as resolved, (i) Ignore incoming addition,
        (m) Merge the directories, (h) Help, (q) Quit resolution: 
```

There's probably a way around that, but it might take a while, given my lack of SVN skills 🤣

mindctrl replied:
@kienstra yeah I think for now let's leave it as is. We'll have to update some other parts of the script to get this in.

mindctrl reviewed (commented):
This looks good to me. Once you add in the `svn ci` command I think it'll be ready to merge. Next week when we do the release we'll be sure to have a backup plan in case the script fails.

kienstra committed:
Add the svn ci command, mainly copied from Genesis Blocks

kienstra commented:
Thanks, @mindctrl! https://github.com/studiopress/genesis-custom-blocks/pull/32/commits/4b92cdd3c1173d16e8536df2d22393c448e9b2c4 adds the `svn ci` command.
